### PR TITLE
Fix dag mutations

### PR DESCRIPTION
**Pull request by vangthao95**
There was an issue with the sequential heuristic list scheduler and LLVM mutations in the second pass. This fix postpones adding the artificial edges from the mutations until after the sequential heuristic list scheduler is finished.

### CODE REVIEW

File: lib/Scheduler/sched_basic_data.cpp [hunk 1]
```diff
@@ -350,7 +350,8 @@ SchedInstruction *SchedInstruction::GetNxtPrdcsr(InstCount *scsrNum,
 
 SchedInstruction *SchedInstruction::GetFrstScsr(InstCount *prdcsrNum,
                                                 UDT_GLABEL *ltncy,
-                                                DependenceType *depType) {
+                                                DependenceType *depType,
+                                                bool *IsArtificial) {
```
Comment (kerbowa) [inline lib/Scheduler/sched_basic_data.cpp:354]: Pass by reference?

File: lib/Scheduler/data_dep.cpp [hunk 2]
```diff
@@ -2972,8 +2973,15 @@ bool InstSchedule::VerifyDataDeps_(DataDepGraph *dataDepGraph) {
...
     UDT_GLABEL ltncy;
     DependenceType depType;
-    for (SchedInstruction *scsr = inst->GetFrstScsr(NULL, &ltncy, &depType);
-         scsr != NULL; scsr = inst->GetNxtScsr(NULL, &ltncy, &depType)) {
+    bool IsArtificial;
+    for (SchedInstruction *scsr =
+             inst->GetFrstScsr(NULL, &ltncy, &depType, &IsArtificial);
+         scsr != NULL;
+         scsr = inst->GetNxtScsr(NULL, &ltncy, &depType, &IsArtificial)) {
+      // Artificial nodes are not required for the schedule to be correct
```
Comment (kerbowa) [inline lib/Scheduler/data_dep.cpp:2981]: Artificial edges

**Comment by vangthao95** [comment]
> Is the idea to always ignore artificial edges? I thought we needed to do this conditionally during sequential scheduling, or am I misunderstanding the idea?

We only ignore artificial edges during sequential scheduling and when verifying the schedule for correctness. One issue I was having was that when we ignore artificial edges during sequential scheduling and then verify the schedule with artificial edges, it will deem the schedule as invalid. This happens if the B&B enumerator was unable to find a schedule with artificial edges enabled. Artificial edges are not required for correctness and this was just my quick fix. If there are any issues with this, please let me know.

File: lib/Scheduler/sched_basic_data.cpp (hunk 1 above)
Comment (vangthao95) [reply lib/Scheduler/sched_basic_data.cpp:354]: I copied the way it was done to retrieve an optional value from a GraphEdge. Is there an issue with this? Although I do agree we can definitely use a better method to retrieve these values.

**Comment by Quincunx271** [comment]
FYI, you don't need to manually run clang-format. I set it up so that GitHub will do it for you if you add a comment with exactly the text: `Do: Reformat`

**Comment by vangthao95** [comment]
> FYI, you don't need to manually run clang-format. I set it up so that GitHub will do it for you if you add a comment with exactly the text: `Do: Reformat`

Do I just put it in the description for the commit?

**Comment by Quincunx271** [comment]
No, you just comment on the PR (this comment box) and your commits will be reformatted.

**Comment by vangthao95** [comment]
Gotcha. Will it also update future commits?

**Comment by Quincunx271** [comment]
No. It just rewrites the existing commits. Adding another comment in the future will update any added commits. You can delete the comment after the clang-format run has completed.

File: lib/Scheduler/sched_region.cpp [hunk 3]
```diff
@@ -244,6 +257,10 @@ FUNC_RESULT SchedRegion::FindOptimalSchedule(
 #endif
   }
 
+  // Log the lower bound on the cost, allowing tools reading the log to compare
+  // absolute rather than relative costs.
+  Logger::Info("Lower bound of cost before scheduling: %d", costLwrBound_);
```
Comment (Quincunx271) [inline lib/Scheduler/sched_region.cpp:262]: Is there a reason why this was moved away from the lower bound cost calculation as opposed to being directly after it?

Comment (vangthao95) [reply lib/Scheduler/sched_region.cpp:262]: No, I must have accidentally placed it too far when I was making this change. There's should be no reason as to why this is so far below the lower bound calculation itself.

Comment (Quincunx271) [reply lib/Scheduler/sched_region.cpp:262]: Okay, good to know.

**Pull request merged**